File: 77-Combinations/77-Combinations.java
/*

77. Combinations
Medium

Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

Example:

Input: n = 4, k = 2
Output:
[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]

*/

class Solution {
    public List<List<Integer>> combine(int n, int k) {
        int[] nums = new int[n];
        for(int i = 0; i < nums.length; i++)
            nums[i] = i + 1;
        List<List<Integer>> resList = new ArrayList<>();
        dfs(k, nums, 0, resList, new ArrayList<>());
        return resList;
    }
    private void dfs(int k, int[]nums,int start, List<List<Integer>> resList, List<Integer> list){
        if(list.size() == k){
            resList.add(new ArrayList<>(list));
        }
        for(int i = start; i < nums.length; i++){
            list.add(nums[i]);
            dfs(k, nums, i + 1, resList, list);
            list.remove(list.size() - 1);
        }
    }
}

File: 77-Combinations/77-Combinations.py
"""

77. Combinations

Given two integers n and k, return all possible combinations of k numbers chosen from the range [1, n].

You may return the answer in any order.

Example 1:

Input: n = 4, k = 2
Output: [[1,2],[1,3],[1,4],[2,3],[2,4],[3,4]]
Explanation: There are 4 choose 2 = 6 total combinations.
Note that combinations are unordered, i.e., [1,2] and [2,1] are considered to be the same combination.

Example 2:

Input: n = 1, k = 1
Output: [[1]]
Explanation: There is 1 choose 1 = 1 total combination.

Constraints:

    1 <= n <= 20
    1 <= k <= n

"""

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        def dfs(idx, temp):
            if len(temp) == k:
                res.append(temp[:])
                return
            for i in range(idx, len(nums)):
                dfs(i + 1, temp + [nums[i]])
        nums = [i for i in range(1, n + 1)]
        dfs(0, [])
        return res
